lib/em/pure_ruby.rb

```diff
@@ -159,10 +160,73 @@ def set_timer_quantum interval
     def epoll
     end
 
-    # This method is not implemented for pure-Ruby implementation
     # @private
     def ssl?
-      false
+      true
+    end
+
+    # This method takes a series of positional arguments for specifying such
+    # things as private keys and certificate chains. It's expected that the
+    # parameter list will grow as we add more supported features. ALL of these
+    # parameters are optional, and can be specified as empty or nil strings.
+    # @private
+    def set_tls_parms signature, priv_key, cert_chain, verify_peer, fail_if_no_peer_cert, sni_hostname, cipher_list, ecdh_curve, dhparam, protocols_bitmask
+      unless sni_hostname.nil? || sni_hostname.empty?
+        raise "set_tls_params sni_hostname has not been implemented in pure Ruby"
+      end
+      unless ecdh_curve.nil? || ecdh_curve.empty?
+        raise "set_tls_params ecdh_curve has not been implemented in pure Ruby"
+      end
+      unless dhparam.nil? || dhparam.empty?
+        raise "set_tls_params dhparam has not been implemented in pure Ruby"
+      end
+      ssl_version, _ = {
+        :TLSv1_2 => EM_PROTO_TLSv1_2,
+        :TLSv1_1 => EM_PROTO_TLSv1_1,
+        :TLSv1 => EM_PROTO_TLSv1,
+        :SSLv3 => EM_PROTO_SSLv3,
+        :SSLv23 => EM_PROTO_SSLv2
+      }.detect do |_, bits|
+        (bits & protocols_bitmask) != 0
+      end
+      @tls_parms ||= {}
+      @tls_parms[signature] = {
+        :verify_peer => verify_peer,
+        :fail_if_no_peer_cert => fail_if_no_peer_cert,
+        :cipher_list => cipher_list,
+        :ssl_version => ssl_version
+      }
+      @tls_parms[signature][:priv_key] = File.read(priv_key) unless priv_key.nil? || priv_key.empty?
+      @tls_parms[signature][:cert_chain] = File.read(cert_chain) unless cert_chain.nil? || cert_chain.empty?
+    end
+
+    def start_tls signature
+      selectable = Reactor.instance.get_selectable(signature) or raise "unknown io selectable for start_tls"
+      tls_parms = @tls_parms[signature]
+      ctx = OpenSSL::SSL::SSLContext.new(tls_parms[:ssl_version])
+      ctx.cert = DefaultCertificate.cert
+      ctx.key = DefaultCertificate.key
+      ctx.cert_store = OpenSSL::X509::Store.new
+      ctx.cert_store.set_default_paths
+      ctx.cert = OpenSSL::X509::Certificate.new(tls_parms[:cert_chain]) if tls_parms[:cert_chain]
+      ctx.key = OpenSSL::PKey::RSA.new(tls_parms[:priv_key]) if tls_parms[:priv_key]
+      verify_mode = OpenSSL::SSL::VERIFY_NONE
+      if tls_parms[:verify_peer] || tls_parms[:verify_peer] == 1
```

Comment:
I'll happily remove the extra conditional, wasn't sure if all consuming libraries were following the rules and using a boolean.